### PR TITLE
Get right most child

### PR DESCRIPTION
**Pull request by areshand**
## Description

When restoring sharded db is in progress, we may haven't write the root node of the sharded JMT when being interrupted.
we still want to get the rightmost child in this case to continue restoring.

## How Has This Been Tested?

added unit test

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Move Compiler
- [ ] Other (specify)

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: storage/aptosdb/src/state_merkle_db.rs [hunk 1]
```diff
@@ -759,9 +744,7 @@ impl StateMerkleDb {
             iter.seek_for_prev(&seek_key)?;
 
             if let Some((node_key, node)) = iter.next().transpose()? {
-                if node_key.version() != version {
-                    continue;
-                }
+                assert_eq!(node_key.version(), version);
```
Comment (msmouse) [inline storage/aptosdb/src/state_merkle_db.rs:747]: let's return error

File: storage/aptosdb/src/state_store/state_store_test.rs [hunk 2]
```diff
@@ -465,6 +465,59 @@ proptest! {
...
+            .get_value_range_proof(rightmost_of_batch1, version)
+            .unwrap();
+
+        restore.add_chunk(batch1, proof_of_batch1).unwrap();
+        restore.wait_for_async_commit().unwrap();
+
+        let _ = store2.state_merkle_db.get_rightmost_leaf_naive(version).unwrap();
+        // When re-initializing the store, the rightmost leaf should exist indicating the progress
+        let res = store2.state_merkle_db.get_rightmost_leaf(version).unwrap();
+        assert!(res.is_some());
```
Comment (msmouse) [inline storage/aptosdb/src/state_store/state_store_test.rs:518]: assert the actual result matches the naive result ?

**Review by grao1991 — APPROVED** [review]

File: storage/aptosdb/proptest-regressions/state_store/state_store_test.txt
Comment (msmouse) [inline storage/aptosdb/proptest-regressions/state_store/state_store_test.txt:1]: remove? (we generally don't commit these)

File: storage/aptosdb/src/state_store/state_store_test.rs [hunk 3]
```diff
@@ -465,6 +465,55 @@ proptest! {
...
+
+        restore.add_chunk(batch1, proof_of_batch1).unwrap();
+        restore.wait_for_async_commit().unwrap();
+
+        let expected = store2.state_merkle_db.get_rightmost_leaf_naive(version).unwrap();
+        // When re-initializing the store, the rightmost leaf should exist indicating the progress
+        let actual = store2.state_merkle_db.get_rightmost_leaf(version).unwrap();
+        // ensure the rightmost leaf is not None
+        assert!(actual.is_some());
+        prop_assert_eq!(actual, expected);
```
Comment (msmouse) [inline storage/aptosdb/src/state_store/state_store_test.rs:514]: using either `assert!` or `prop_assert_eq!` is fine, preferring `prop_assert_eq!` maybe, but a mix is kinda confusing..

Comment (areshand) [reply storage/aptosdb/src/state_store/state_store_test.rs:514]: just want to explicitly test `actual` exists, which is the bug we found. I feel assert eq as you suggested is not enough as expected can be None

**Review by msmouse — APPROVED** [review]

**Pull request merged**